core/src/test/scala/unit/kafka/server/HeartbeatRequestTest.scala

```diff
@@ -0,0 +1,145 @@
+    new ClusterConfigProperty(key = "group.coordinator.new.enable", value = "false"),
+    new ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "1"),
+    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1")
+  ))
+  def testHeartbeatWithOldConsumerGroupProtocolAndOldGroupCoordinator(): Unit = {
+    testHeartbeat()
+  }
+
+  private def testHeartbeat(): Unit = {
+
+    // Creates the __consumer_offsets topics because it won't be created automatically
+    // in this test because it does not use FindCoordinator API.
+    createOffsetsTopic()
+
+    // Create the topic.
+    createTopic(
+      topic = "foo",
+      numPartitions = 3
+    )
+
+    for (version <- ApiKeys.HEARTBEAT.oldestVersion() to ApiKeys.HEARTBEAT.latestVersion(isUnstableApiEnabled)) {
+
+      val (memberId, memberEpoch) = joinDynamicConsumerGroupWithOldProtocol(
+        groupId = "grp",
+        completeRebalance = false
+      )
+
+      // Heartbeat with unknown group id and unknown member id.
+      heartbeat(
+        groupId = "grp-unknown",
+        memberId = "member-id-unknown",
+        generationId = -1,
+        expectedError = Errors.UNKNOWN_MEMBER_ID,
+        version = version.toShort
+      )
+
+      // Heartbeat with unknown group id.
+      heartbeat(
+        groupId = "grp-unknown",
+        memberId = memberId,
+        generationId = -1,
+        expectedError = Errors.UNKNOWN_MEMBER_ID,
+        version = version.toShort
+      )
+
+      // Heartbeat with unknown member id.
+      heartbeat(
+        groupId = "grp",
+        memberId = "member-id-unknown",
+        generationId = -1,
+        expectedError = Errors.UNKNOWN_MEMBER_ID,
+        version = version.toShort
+      )
+
+      // Heartbeat with unmatched generation id.
+      heartbeat(
+        groupId = "grp",
+        memberId = memberId,
+        generationId = -1,
+        expectedError = Errors.ILLEGAL_GENERATION,
+        version = version.toShort
+      )
+
+      // Heartbeat COMPLETING_REBALANCE group.
+      heartbeat(
+        groupId = "grp",
+        memberId = memberId,
+        generationId = memberEpoch,
+        version = version.toShort
+      )
+
+      syncGroupWithOldProtocol(
+        groupId = "grp",
+        memberId = memberId,
+        generationId = memberEpoch
+      )
+
+      // Heartbeat STABLE group.
+      heartbeat(
+        groupId = "grp",
+        memberId = memberId,
+        generationId = memberEpoch,
+        version = version.toShort
+      )
+
+      leaveGroup(
+        groupId = "grp",
+        memberId = memberId,
+        useNewProtocol = false,
+        version = ApiKeys.LEAVE_GROUP.latestVersion(isUnstableApiEnabled)
+      )
+
+      // Heartbeat empty group.
+      heartbeat(
+        groupId = "grp",
+        memberId = memberId,
+        generationId = -1,
+        expectedError = Errors.UNKNOWN_MEMBER_ID,
+        version = version.toShort
+      )
+    }
```

Comment:
Have you considered adding one while in PREPARING_REBALANCE state in order to verify that we get the REBALANCE_IN_PROGRESS error?